core/modules/qdisp/Executive.cc

```diff
@@ -150,61 +119,52 @@ Executive::Executive(Config::Ptr c, std::shared_ptr<MessageStore> ms)
 
 Executive::~Executive() {
     // Real XrdSsiService objects are unowned, but mocks are allocated in _setup.
-    delete dynamic_cast<XrdSsiServiceMock *>(_service);
+    delete dynamic_cast<XrdSsiServiceMock *>(_xrdSsiService);
 }
 
-class Executive::DispatchAction : public util::VoidCallable<void> {
-public:
-    typedef std::shared_ptr<DispatchAction> Ptr;
-    DispatchAction(Executive& executive,
-                   int jobId,
-                   Executive::JobDescription const& jobDescription,
-                   JobStatus::Ptr jobStatus)
-        : _executive(executive), _jobId(jobId),
-          _jobDescription(jobDescription), _jobStatus(jobStatus) {
-    }
-    virtual ~DispatchAction() {}
-    virtual void operator()() {
-        if(_jobDescription.requester->reset()) { // Must be able to reset state
-            _executive._dispatchQuery(_jobId, _jobDescription, _jobStatus);
-        }
-        // If the reset fails, do nothing-- can't retry.
-    }
-private:
-    Executive& _executive;
-    int _jobId;
-    JobDescription _jobDescription;
-    JobStatus::Ptr _jobStatus; ///< Points at status in Executive::_statusMap
-};
-
-/* Add a new job to executive queue, if not already in. Not thread-safe.
+/** Add a new job to executive queue, if not already in. Not thread-safe.
  */
-void Executive::add(int jobId, Executive::JobDescription const& jobDescription) {
-    bool alreadyCancelled = lockedRead(_cancelled, _cancelledMutex);
-    if(alreadyCancelled) {
-        LOGF(getLogger(), LOG_LVL_INFO, "Executive already cancelled, ignoring add(%1%)" % jobId);
+void Executive::add(JobDescription const& jobDesc) {
+    LOGF_DEBUG("Executive::add(%1%)" % jobDesc.toString());
+    if(_cancelled.get()) {
+        LOGF(getLogger(), LOG_LVL_INFO, "Executive already cancelled, ignoring add(%1%)" % jobDesc.id());
         return;
     }
-    bool trackOk = _track(jobId, jobDescription.requester); // Remember so we can join
-    if(!trackOk) {
-        LOGF(getLogger(), LOG_LVL_WARN, "Ignoring duplicate add(%1%)" % jobId);
+    // Create the JobQuery and put it in the map.
+    JobStatus::Ptr jobStatus = std::make_shared<JobStatus>();
+    MarkCompleteFunc::Ptr mcf = std::make_shared<MarkCompleteFunc>(this, jobDesc.id());
+    JobQuery::Ptr jobQuery(new JobQuery(this, jobDesc, jobStatus, mcf));
+    jobQuery->setupWeakThis(jobQuery);
+    if(!_addJobToMap(jobQuery)) {
+        LOGF_ERROR("Executive ignoring duplicate job add(%1%)" % jobQuery->getId());
```

Comment:
Why do you prefer to log to global/root logger instead of the logger returned by `getLogger()` defined here? We should try to keep it consistent in one file.